guides/v2.0/extension-dev-guide/proxies.md

```diff
@@ -0,0 +1,92 @@
+
+If a class's constructor is particularly resource-intensive, this can lead to unnecessary performance impact when another class depends on it, if the expensive object does not end up being needed during a particular request. (You can display a *dependency graph* of such objects by enabling <a href="{{ site.gdeurl }}config-guide/bootstrap/mage-profiler.html">profiling</a>.)
+
+As an example, consider the following two classes:
+
+{% highlight PHP %}
+<?php
+class SlowLoading
+{
+    public function __construct()
+    {
+        // ... Do something resource intensive
+    }
+
+    public function getValue()
+    {
+        return 'SlowLoading value';
+    }
+}
+
+class FastLoading
+{
+    protected $slowLoading;
+
+    public function __construct(
+        SlowLoading $slowLoading
+    ){
+        $this->slowLoading = slowLoading;
+    }
+
+    public function getFastValue()
+    {
+        return 'FastLoading value';
+    }
+
+    public function getSlowValue()
+    {
+        return $this->slowLoading->getValue();
+    }
+}
+?>
+{% endhighlight %}
+
+Assume that class `SlowLoading` has a non-trivial performance impact when instantiated (perhaps due to a complex database query or a call to a third-party web API). Because of the dependency injection in the constructor of `FastLoading`, this impact is incurred if `FastLoading` is instantiated.  Note, however, that the `SlowLoading` instance is used only in the method `getSlowValue`, meaning that the resource cost is unnecessary if this method is never called on the `FastLoading` object.
+
+### Proxies are generated code
+Magento has a solution for this situation: proxies. <a href="http://en.wikipedia.org/wiki/Proxy_pattern" target="_blank">Proxies</a> extend other classes to become lazy-loaded versions of them. That is, a real instance of the class a proxy extends created only after one of the class's methods is actually called. A proxy implements the same interface as the original class and so can be used as a dependency anywhere the original class can.  Unlike its parent, a proxy has only once dependency: the object manager.
+
+Proxies are generated code and therefore do not need to be manually written.  (See <a href="{{ site.gdeurl }}extension-dev-guide/code-generation.html">Code generation</a> for more information.) Simply reference a class in the form `\Original\Class\Name\Proxy`, and the class is generated if it does not exist.
+
+### Solve the issue using a proxy
+Using the preceding example, the constructor signature of FastLoading could be changed as follows:
+
+{% highlight PHP %}
+<?php
+    public function __construct(
+        SlowLoading\Proxy $slowLoading
```

Comment:
Fair point.  I went ahead and modified this.  Since this was a larger text change, I made the modification in the original branch of the Classy Llama fork and submitted a PR to this branch instead of 2.0.